Introduce Roslyn source generators for Sync method generation

penev92 opened the pull request:
Inspired by @teinarss'es wip branch and articles like https://andrewlock.net/exploring-dotnet-6-part-9-source-generator-updates-incremental-generators, this aims to bring a mostly elegant solution to #21032.


_**This is a pretty serious status quo change so naturally everything here is open for discussion and everyone should pitch in!**_


Things to note:
- Projects containing source generators must target `netstandard2.0` and that is a fact of life.
- Source generators work by abusing partial classes, so we must accept that if we are to use them.
~~- The convention is "if a class is partial and implements `ISync`, then we should generate an implementation for it".~~  **[Obsolete point after the PR was rewritten.]** Now replaced with a `GenerateSyncCodeAttribute` (inheritable). Classes just need to be partial and have or inherit the attribute.
~~- I have actually found a use for the relatively new to C# interface default method implementations with the new `ISync.GetSyncHash()`. Throwing by default means we are notified if a sync function for a type isn't generated.~~ **[Obsolete point after the PR was rewritten.]**
- Some hashing values (like for Actor, Player and bool) are weird.
- Despite all the previous trait and sync function caching, this should allow us to do a lot less when calculating sync hashes.
- The performance gains seem to be real, but situational.
- **[NEW]** After the PR was rewritten, we should not be syncing *a lot* less traits than we do on bleed.
- As always, reviewing by commit will ***help***.

Performance benchmarks should be done (by others)! I've personally done a few and they show a significant decrease in time spent inside `OpenRA.World.SyncHash()`, which I can only interpret as better overall performance (ofc mostly felt when there are *lots* of actors on the map with lots of traits to sync). Big AI team games showed 60%+ of total game CPU time spent inside that method on bleed and 40%+ on this branch. Take that information with a cup of salt and do with it as you like. _(At the same time, on another, obviously GPU-bound computer, the same method went from taking 4.3% of CPU time to 3.7%, while 60-80% was spent on rendering, so again - this is a very situational performance boost)_

![image](https://github.com/OpenRA/OpenRA/assets/7137365/90f6d646-3e7a-47de-bae7-1b063f6814a5)

Closes #21032.

penev92 committed:
Disabled SA1205 - PartialElementsMustDeclareAccess

Otherwise partial types require an access modifier, but:
- Public sealed partial traits make no sense.
- Internal sealed partial traits are forbidden by our conventions (can't explicitly use default modifiers).
- We prefer sealed over unsealed traits (a point of contestion, so we keep the status quo).

pchote commented on OpenRA.SourceGenerators/SourceGeneratorHelpers.cs:
```diff
@@ -0,0 +1,41 @@
+﻿using System;
```
Copyright header?

pchote commented on OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,95 @@
+﻿using System.Collections.Generic;
```
and here.

pchote commented on OpenRA.SourceGenerators/Sync/SyncHelpers.cs:
```diff
@@ -0,0 +1,62 @@
+﻿using System;
```
^

penev92 commented:
This now passes the CI checks and is (supposedly) feature-complete.
However, https://github.com/dotnet/vscode-csharp/issues/6426 will likely prove to be a **_blocker_** as VSCode with the C# extension is currently completely unable to navigate to generated code and code related to it and I don't want to make the developer experience worse for anyone.
Aside from that, there are a couple of things to do while we wait:
- Discuss the hashing functions for individual types
- Run more diverse benchmarks
- Check with more developers that their IDEs are fine with this

RoosterDragon commented on OpenRA.Game/OpenRA.Game.csproj:
```diff
@@ -15,5 +15,8 @@
     <PackageReference Include="System.Threading.Channels" Version="6.0.0" />
+	<ProjectReference Include="..\OpenRA.SourceGenerators\OpenRA.SourceGenerators.csproj"
+					  OutputItemType="Analyzer"
+					  ReferenceOutputAssembly="false" />
```
Review all changed .csproj files as they have a mix of space and tab indents.

RoosterDragon replied:
Still mixed indents.

RoosterDragon commented on OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,106 @@
+				return (null, null);
+
+			var semanticModel = compilation.GetSemanticModel(classDeclaration.SyntaxTree);
+			var typeSymbol = semanticModel.GetDeclaredSymbol(classDeclaration) as ITypeSymbol;
```
Why this cast to ` as ITypeSymbol`? `GetDeclaredSymbol` returns `INamedTypeSymbol` which dervies from `ITypeSymbol` anyway. Getting rid of it seems harmless, and also means we don't need the extenion method `GetAllMembers` on `ITypeSymbol`.

If we do need this cast for some reason, it should be a direct cast for correctness.

RoosterDragon commented on OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,106 @@
+namespace {namespaceName}
+{{
+	// Explicitly implement the interface again to avoid issues if it is inherited.
+	{(isSealed ? "sealed" : "public")} partial class {className} : {SyncHelpers.SyncInterfaceName}
```
This `isSealed` check feels a bit weak, what if we have a public sealed trait for example? Do we need something that tries to maintain all the keyword modifiers applied to the class instead?

RoosterDragon commented on OpenRA.Roslyn/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+      "targetProject": "..\\OpenRA.Mods.Common\\OpenRA.Mods.Common.csproj"
+    }
+  }
+}
```
Visual Studio doesn't like this profile:
![image](https://github.com/OpenRA/OpenRA/assets/3399086/e2c3e27e-b468-49f4-9e5f-e01eaee80020)

RoosterDragon replied:
It's still not a fan of the revised profiles. Is this something VS just won't gel with?

teinarss replied:
Found this, https://github.com/JoanComasFdz/dotnet-how-to-debug-source-generator-vs2022 so you will need to install .NET Compiler Platform SDK.  see https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/

RoosterDragon commented on OpenRA.Roslyn/SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,106 @@
+			return {string.Join(" ^ ", syncMembersAsHashCodeStrings)};
+		}}{(syncsTargets ? $"\n{SyncHelpers.GetTargetSyncHashMethod()}" : string.Empty)}
+	}}
+}}");
```
This generates code with all the hashing effectively inlined in source.

```c#
// <auto-generated/>
using OpenRA.Traits;

namespace OpenRA.Mods.Common.Traits
{
	// Explicitly implement the interface again to avoid issues if it is inherited.
	public partial class AttackBomber : OpenRA.ISync
	{
		int ISync.GetSyncHash()
		{
			return HashTarget(target) ^ (inAttackRange ? 0xaaa : 0x555) ^ (facingTarget ? 0xaaa : 0x555) ^ (IsAiming ? 0xaaa : 0x555) ^ (IsTraitPaused ? 0xaaa : 0x555) ^ (IsTraitDisabled ? 0xaaa : 0x555);
		}

		static int HashTarget(Target t)
		{
			switch (t.Type)
			{
				case TargetType.Actor:
					return (t.Actor == null ? 0 : (int)(t.Actor.ActorID << 16));

				case TargetType.FrozenActor:
					var actor = t.FrozenActor.Actor;
					if (actor == null)
						return 0;

					return (actor == null ? 0 : (int)(actor.ActorID << 16));

				case TargetType.Terrain:
					return t.CenterPosition.X ^ t.CenterPosition.Y ^ t.CenterPosition.Z;

				case TargetType.Invalid:
				default:
					return 0;
			}
		}
	}
}
```

I would prefer to keep `Sync.HashTarget` and use that rather than copy-pasting it into every use site.

Keeping all the Sync helpers could be somewhat nice, e.g. if we have `Sync.Hash(Target target)`, `Sync.Hash(CPos cell)`, etc then the generated code looks like:
```c#
int ISync.GetSyncHash()
{
			return Sync.Hash(target) ^ Sync.Hash(inAttackRange) ^ Sync.Hash(facingTarget);
}
```

This also means we can revert `Added GetSyncHash() method to ISync, Actor, Player` which I don't like as a commit. Keeping all the sync helpers in `Sync` means we centralise `Sync` related things in one place. Moving `GetSyncHash` to individual classes means spreading out that knowledge. It's much nicer to have all the helpers in Sync - we can easily see what is and isn't supported and change how they work, rather than having things scattered about.

If we wanted, we could even define some `int Sync.CombineHash(int a, int b, int c, int d) => a ^ b ^ c ^ d` overloads that allows us to do
```c#
int ISync.GetSyncHash()
{
			return Sync.CombineHash(Sync.Hash(target), Sync.Hash(inAttackRange), Sync.Hash(facingTarget));
}
```

This allows us to keep the helpers in non-generated code which is somewhat more accessible for modifications imo. The JIT can dealing with inlining everything.

If the `Sync.Hash` helpers are used, then then source generator doesn't even need to know what types are supported by the sync attribute. It just calls the helper. We could define `Sync.Hash(object _) => throw new NotSupportException("with a descriptive error");` to detect `[Sync]` on unsupported members. Again I think this would help keep things more accessible as the source generator doesn't have to know as much about the domain.

RoosterDragon commented on OpenRA.SourceGenerators/Sync/SyncHelpers.cs:
```diff
@@ -0,0 +1,73 @@
+				"Actor" => $"({objectName} == null ? 0 : (int)({objectName}.ActorID << 16))",
+				"Player" => $"({objectName} == null ? 0 : (int)({objectName}.PlayerActor.ActorID << 16) * 0x567)",
+				"Target" => $"HashTarget({objectName})",
+				_ => throw new NotSupportedException()
```
If I define `[Sync] long Field;` this hits the `throw new NotSupportedException()` in the source generator and I get a warning in the IDE. This exception should provide a descriptive error message on the problem, how to fix it, and a list of supported types for good dev experience.

Alternatively if we go down the `Sync.Hash(object _) => throw new NotSupportedException()` then that needs to have such a message instead.

RoosterDragon commented on OpenRA.Game/Actor.cs:
```diff
@@ -154,7 +146,7 @@ internal Actor(World world, string name, TypeDictionary initDict)
 				var targetablesList = new List<ITargetable>();
 				var targetablePositionsList = new List<ITargetablePositions>();
-				var syncHashesList = new List<SyncHash>();
+				var syncTraitsList = new List<ISync>();
```
Nit: `syncsList` to match existing naming convention.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Player/MissionObjectives.cs:
```diff
@@ -89,7 +89,7 @@ public int ObjectivesHash
 				var hash = 0;
 				foreach (var objective in objectives)
-					hash ^= Sync.HashUsingHashCode(objective.State);
+					hash ^= objective.State.GetHashCode();
```
~~This isn't safe, we need to use Sync's hashing mechanism, which we know is deterministic across processes. .NET's hashing doesn't make those same guarentees.~~

Can we cast to int directly here instead of calling GetHashCode?

RoosterDragon commented on OpenRA.Game/Sync.cs:
```diff
@@ -10,158 +10,21 @@
+	[RequireExplicitImplementation]
+	public interface ISync
+	{
+		int GetSyncHash() => throw new NotImplementedException();
```
For good dev experience, this should provide a detailed error about how to implement ISync correctly. Either that you can manually implement it, or that you can tag supported types with `[Sync]` and make the class partial to get an implementation provided automatically.

penev92 replied:
This no longer has a default implementation so I don't know what we can do about this comment. Add a `<summary>` on the interface?

RoosterDragon replied:
Yes, some comments on the interface(s) to explain intended usage would be fine by me.

RoosterDragon commented on OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,106 @@
+			var syncedClassMembers = typeSymbol.GetAllMembers(x => x.HasSyncAttribute()).ToArray();
+			var hashCodeStrings = new List<string>();
+			var syncsTargets = false;
+			foreach (var member in syncedClassMembers)
```
Dev experience thought: If you haven't tagged any members then maybe we should throw an error to make the dev aware. This could be wrapped in a helper, e.g.
```c#
int ISync.GetSyncHash()
{
    throw Sync.NoSyncMembers();
}
```
```c#
NotSupportedException NoSyncMembers()
{
    return new NotSupportedException("descriptive error about using sync tag here");
}
```

penev92 replied:
I believe the lint rules should already handle that?
Turning it into a proper analyzer sounds fun though.

RoosterDragon commented on OpenRA.Mods.Common/Traits/AffectsShroud.cs:
```diff
@@ -41,10 +41,10 @@ public abstract class AffectsShroud : ConditionalTrait<AffectsShroudInfo>, ISync
 
 		[Sync]
-		WDist cachedRange;
+		protected WDist CachedRange { get; private set; }
```
Hmmm - the old IL emit was allowed to access private types - but the source generator plays by the rules so you can't go accessing private types in a base class. If you try and do this, then the generated code raises the usual C# error about an inaccessible member. So it's a smooth experience and should be easy for people to fix...

... unless you don't have access to the base class. If you're a mod dll and the private sync'd field is in a common trait, then you can't get at it to fix the error.

I think we need to apply some thought as to how sync works across dervied classes. The old IL gen could just access the private fields and didn't care. But the source gen is going to get in a twist about that.

Maybe we need to implicitly implement the interface as `virtual` so derived classes can chain to it? `int GetSyncHash() => SomeField ^ base.GetSyncHash();`

pchote replied:
Would it be feasible (perf-wise) for the generated code add logic to the constructor to fetch using reflection and create a delegate to access private base fields?

RoosterDragon replied:
I suspect the perf wins from this PR are from avoiding the indirection of invoking a delegate. Would need a benchmark to be sure but overall I imagine being able to invoke the interface method (even though that's virtual) gives more info the to JIT than a delegate that black boxes everything.

penev92 replied:
Yes, removing the delegates was one of my main goals and I expect you're right about most of the perf gain coming from there.
I went with each class hashing its members and calling the base class'es `GetSyncHash()` so we don't have to resort to black magic in order to access private members.

RoosterDragon commented on OpenRA.SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,106 @@
+
+			var isSealed = classDeclaration.IsSealed();
+
+			return GenerateClass(namespaceDeclaration.Name, classDeclaration.Identifier.Text, hashCodeStrings, isSealed, syncsTargets);
```
If my `partial` class already implements `ISync` directly in the type (and not from a base type) then we shouldn't generate code for this, or else C# complains that we're trying to implement the interface twice.

RoosterDragon reviewed (changes requested):
Some initial thoughts for you. I think getting a good dev experience out of this may take some careful planning when working outside the happy path.

penev92 committed:
Switched implementing ISync to GenerateSyncCode

Types that have members with `SyncMemberAttribute`  MUST be partial AND MUST either have or inherit `GenerateSyncCodeAttribute`.

penev92 commented:
I believe I addressed most of the above comments as well as somewhat significantly changed how things are done based on discussions on Discord.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Conditions/ConditionalTrait.cs:
```diff
@@ -49,7 +49,6 @@ public virtual IEnumerable<VariableObserver> GetVariableObservers()
 				yield return new VariableObserver(RequiredConditionsChanged, Info.RequiresCondition.Variables);
 		}
 
-		[Sync]
```
Would PR this commit (`Stopped syncing IsTraitPaused/Disabled`) separately, so that making a fair comparison of the perf changes on the source generators is possible.

RoosterDragon commented on OpenRA.Mods.Cnc/Traits/Attack/AttackPopupTurreted.cs:
```diff
@@ -46,7 +46,7 @@ sealed class AttackPopupTurretedInfo : AttackTurretedInfo, Requires<BuildingInfo
 	}
 
-	sealed class AttackPopupTurreted : AttackTurreted, INotifyIdle, IDamageModifier
+	sealed partial class AttackPopupTurreted : AttackTurreted, INotifyIdle, IDamageModifier
```
There's a lot of files made partial in this commit that don't need to be partial - they don't have any syncable members (only their base classes do).

Similarly there's lot of pointless overrides that just defer to base:
```c#
	public partial class AttackTurreted : OpenRA.ISync
	{
#pragma warning disable CS1591 // Apparently disabling in .editorconfig doesn't work for generated code.
		public override int GetSyncHash()
		{
			return base.GetSyncHash();
		}
#pragma warning restore CS1591
	}
```

There's no need to generate such a method.

Whilst I have an example here, consider moving `#pragma warning disable CS1591 ...` to the top of the file and not restoring it. Makes things slightly cleaner to read.

RoosterDragon commented on OpenRA.Game/Sync.cs:
```diff
@@ -10,158 +10,68 @@
-				return (int)(p.PlayerActor.ActorID << 16) * 0x567;
-			return 0;
-		}
+		public static int Hash(object o) => throw new NotSupportedException($"Type {o.GetType().FullName} is not supported for Sync hashing!");
```
I realised we might be able to bin this overload and move this to compile time if the source generator can detect that there is no matching overload and emit something like:
```c#
public virtual int GetSyncHash()
{
	#error "Type <foo> is not supported for Sync hashing on member <bar>"
	return 0;
}
```

PunkPun commented:
related #19535

penev92 commented:
### We are still blocked, waiting on
https://github.com/dotnet/vscode-csharp/issues/6426
https://github.com/dotnet/vscode-csharp/pull/5858
https://github.com/dotnet/roslyn/pull/68771

penev92 committed:
Added a source generator for sync hashing

This will supersede the current approach of creating hashing methods at runtime via IL weaving.
Also added GenerateSyncCodeAttribute and ISync.GetSyncHash() and cleaned up Sync's hashing methods.

penev92 commented:
Split some things off in #21823.

penev92 commented:
The prerequisites are merged, this is rebased and ready for another round.
I addressed whatever comments I could except
> It's still not a fan of the revised profiles. Is this something VS just won't gel with?

because I'm not sure what's wrong with them and
> we might be able to bin this overload and move this to compile time if the source generator can detect that there is no matching overload and emit something like

which sounds cool but I couldn't think of a non-painful way to do it and I feel CodeLens does a pretty good job of telling you *where* you messed up if you hit that case. 🤷‍♂️

RoosterDragon commented on OpenRA.Roslyn/SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,119 @@
+{{
+	{classDeclaration.Modifiers} class {className} : {SyncHelpers.SyncInterfaceName}
+	{{
+		public {(shouldCallBase ? "override" : (isSealed ? "" : "virtual"))} int GetSyncHash()
```
Nitpicking: This generates `public  int GetSyncHash()` with two spaces when neither override/virtual is needed.

RoosterDragon commented on OpenRA.Game/Sync.cs:
```diff
@@ -10,159 +10,84 @@
-				return (int)(p.PlayerActor.ActorID << 16) * 0x567;
-			return 0;
-		}
+		public static int Hash(object o) => throw new NotSupportedException($"Type {o.GetType().FullName} is not supported for Sync hashing!");
```
We could use the `Obsolete` attribute to produce a warning/error if you call the `Hash(object o)` overload.

penev92 replied:
Yeah ok, that was a good idea 👌 
<img width="1133" height="172" alt="image" src="https://github.com/user-attachments/assets/3b02bfbb-f6ac-4492-ae38-9f0c7133289a" />

RoosterDragon commented on OpenRA.Roslyn/SourceGenerators/Sync/SyncHashCodeSourceGenerator.cs:
```diff
@@ -0,0 +1,119 @@
+	{{
+		public {(shouldCallBase ? "override" : (isSealed ? "" : "virtual"))} int GetSyncHash()
+		{{
+			return {string.Join(" ^ ", elements)};
```
I think hiding this mixing function behind a function call would be nicer so we can more easily change this in future. XORing elements together is a crap way to hash lots of things together - though I appreciate it's what we do today.

e.g.
```c#
public static int CombineHashes(int a) => a;
public static int CombineHashes(int a, int b) => a ^ b;
public static int CombineHashes(int a, int b, int c) => a ^ b ^ c;
public static int CombineHashes(int a, int b, int c, int d) => a ^ b ^ c ^ d;
public static int CombineHashes(params ReadOnlySpan<int> hashes)
{
    var ret = 0;
    foreach (var hash in hashes)
        ret ^= hash;
    return ret;
}
```

called as 
```c#
return Sync.CombineHashes(
    Sync.Hash(a),
    Sync.Hash(b)
);

RoosterDragon commented on OpenRA.Roslyn/SourceGenerators/Sync/SyncHelpers.cs:
```diff
@@ -0,0 +1,46 @@
+	static class SyncHelpers
+	{
+		public const string GenerateSyncCodeAttributeName = "GenerateSyncCodeAttribute";
+		public const string VerifySyncAttributeName = "VerifySyncAttribute";
```
Are we able to compare full names for these, on the off chance something conflicts with them from another namespace (e.g. 3rd party mod)?

RoosterDragon commented on OpenRA.Roslyn/OpenRA.Roslyn.csproj:
```diff
@@ -0,0 +1,19 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+		<TargetFramework>netstandard2.0</TargetFramework>
```
Should target net8 like everything else.

penev92 replied:
I'll give it a shot, but I recall back in the day there was something about analyzers needing to target netstandard 🤷‍♂️

penev92 replied:
Seems to work, but I get this warning about unpredictable behaviour 🤷‍♂️ 
> [RS1041](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/rules/RS1041.md): This compiler extension should not be implemented in an assembly with target framework '.NET 8.0'. References to other target frameworks will cause the compiler to behave unpredictably.
<img width="1230" height="210" alt="image" src="https://github.com/user-attachments/assets/b5ce43a1-4155-420c-9b05-efa9d3a948a1" />

RoosterDragon commented on OpenRA.Roslyn/OpenRA.Roslyn.csproj:
```diff
@@ -0,0 +1,19 @@
+		</PackageReference>
+		<!-- Specifically targetting 4.7.0 because the GH Actions worker uses that. -->
+		<!-- And unupdated IDEs will likely not work with newer versions. -->
+		<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.14.0" />
```
Comment no longer aligns with versions here. I wonder if targetting net8 will help if you wanted a newer version anyway?

RoosterDragon replied:
https://github.com/dotnet/roslyn/blob/main/docs/wiki/NuGet-packages.md#versioning suggests maybe ~~`4.11` is a good one to target. Latest net8 without any net9 stuff.~~ `4.8` is a good one to target. Earliest version supporting net8, so not possible for this to be newer than anything people are running locally.

RoosterDragon commented on OpenRA.Game/Sync.cs:
```diff
@@ -10,159 +10,84 @@
+	/// <summary>
+	/// Part of the game state synchronization mechanism.
+	/// Types tagged with this attribute (traits, weapons, effects) MUST be partial classes
+	/// and will get an implementation of ISync generated for them if they don't implement it explicitly.
```
`<see ...` for `ISync `.

RoosterDragon commented on OpenRA.Mods.Cnc/Traits/Attack/AttackTesla.cs:
```diff
@@ -39,7 +39,7 @@ sealed class AttackTeslaInfo : AttackBaseInfo
 	}
 
-	sealed class AttackTesla : AttackBase, ITick, INotifyAttack
+	sealed partial class AttackTesla : AttackBase, ITick, INotifyAttack
```
Shouldn't need `partial` on classes like this one that don't have `[GenerateSyncCode]`, correct?

penev92 replied:
`[GenerateSyncCode]` here is inherited AND an `ISync` implementation is generated because `AttackTesla` syncs `charges` and `timeToRecharge`. (I did remove `partial` from all the types that rely solely on their base type to handle sync generation as you asked previously 😉 )

pchote replied:
Removing `partial` here still compiles without any warnings (even under `make check`), but presumably isn't checked. This is dangerous! Can we add a check, similar to the explicit violations, that looks for classes that have checked fields that aren't checked?

RoosterDragon replied:
I'd also be happy if we used `[GenerateSyncCode]` expliclitly on derived classes when we need it and didn't rely on detecting it from inheritance.

penev92 replied:
`[GenerateSyncCode]` can no longer be inherited, so this is a mute point - every class will need to be explicitly tagged with the attribute, which will cause source generation, which will cause a compile error if the class is not partial.

RoosterDragon reviewed (changes requested):
My IDE experience is certainly a lot nicer than when we last tried this. It's just working in VS 2026. :)

<img width="907" height="285" alt="Image" src="https://github.com/user-attachments/assets/7058f5a8-ae19-48d1-985c-0f820f1ccec9" />

Some assorted comments for you, but overall concept is looking quite good to me here.

RoosterDragon commented on OpenRA.Roslyn/SourceGenerators/Sync/SyncHelpers.cs:
```diff
@@ -0,0 +1,46 @@
+		public static bool ImplementsISync(this INamedTypeSymbol type) => type.AllInterfaces.Any(y => y.ToDisplayString() == SyncInterfaceName);
+
+		public static bool ManuallyImplementsISync(this INamedTypeSymbol type)
+			=> type.ImplementsISync() && type.MemberNames.Any(x => x == SyncHashMethodName);
```
`MemberNames.Contains(SyncHashMethodName)`?

penev92 commented:
Updated the source generator.
Addressed most comments, but we still need to figure out whether we'll actually want to keep the generator post-run.

PunkPun reviewed (commented):
Is this PR waiting on something specific?
